### PR TITLE
Fix deadlock in first catboost model execution

### LINKED ISSUE #13832: First modelEvaluate() call on a newly created CatBoost model deadlocks forever
Hello,

so I'm working to integrate CatBoost model with our ClickHouse cluster. It appears that, whenever we create a new XML model file in the configured path, the **first** call to the new model never returns (this is true both of the HTTP interface as well as via clickhouse-client).

Independently from whether the locked statement gets cancelled, subsequent calls succeed immediately.

```
16e8bdb5bbb9 :) select version();

SELECT version()

┌─version()─┐
│ 20.6.3.28 │
└───────────┘

1 rows in set. Elapsed: 0.006 sec.
```

Is this known behavior? If it helps, my models look like this:

```
<models>
    <model>
        <type>catboost</type>
        <name>ctr_predictor</name>
        <path>/catboost/my_predictor</path>
        <lifetime>0</lifetime>
    </model>
</models>
```

### PR DESCRIPTION
**Pull request by Avogar**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix deadlock in first catboost model execution. Closes #13832

Detailed description / Documentation draft:
.

### CODE REVIEW

File: tests/integration/test_catboost_model_first_evaluate/model/model_config.xml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<models>
+    <model>
+        <type>catboost</type>
+        <name>titanic</name>
+        <path>/etc/clickhouse-server/model/model.bin</path>
```
Comment (alexey-milovidov) [inline tests/integration/test_catboost_model_first_evaluate/model/model_config.xml:5]: Btw, where is `model.bin` file in this PR?

File: src/Interpreters/ExternalLoader.cpp [hunk 2]
```diff
@@ -845,9 +847,10 @@ class ExternalLoader::LoadingDispatcher : private boost::noncopyable
                 if (info.state_id >= min_id)
                     continue;
 
-                all_ready = false;
                 if (info.loading_id < min_id)
                     startLoading(info, forced_to_reload, *min_id);
+
+                all_ready &= (info.state_id >= min_id);
```
Comment (alexey-milovidov) [inline src/Interpreters/ExternalLoader.cpp:853]: It's interesting how the loading of models was different from dictionaries... why there was a deadlock for models and everything was good for dictionaries?

File: tests/integration/test_catboost_model_first_evaluate/model/model_config.xml (hunk 1 above)
Comment (Avogar) [reply tests/integration/test_catboost_model_first_evaluate/model/model_config.xml:5]: It disappeared somehow, don't know why. Will add it.

File: src/Interpreters/ExternalLoader.cpp (hunk 2 above)
Comment (Avogar) [reply src/Interpreters/ExternalLoader.cpp:853]: The problem appears only when `enable_async_loading = true`. In `ExternalDictionariesLoader` it sets to `true` in the constructor, but not in `ExternalModelsLoader`

**Review by alexey-milovidov — APPROVED** [review]

**Comment by Avogar** [comment]
@alexey-milovidov Can we merge it? Failed tests are not related.

**Pull request merged**